### PR TITLE
✨ prebuilt container for clusterctl in image based pipelines

### LINKED ISSUE #5894: Client Container
As a user/operator I would like to use a prebuilt container for clusterctl in image based pipelines.

**Detailed Description**

It would be nice if there was a released official container containing clusterctl along with the released binaries. This could be used in image based pipelines such as gitlab, tekton, etc.

/kind feature

### PR DESCRIPTION
**Pull request by sachinkumarsingh092**
Signed-off-by: Sachin Kumar Singh <sachins2@vmware.com>

**What this PR does / why we need it**:
Add functionality to release an official container for `clusterctl` along with the released binaries to be used in image-based pipelines.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #5894 

/cc @killianmuldoon @fabriziopandini @sbueringer

### CODE REVIEW

File: Dockerfile
Comment (sachinkumarsingh092) [inline Dockerfile:58]: Will it be okay to make this a separate RUN or chain the go build in a single command?

Comment (alexander-demichev) [reply Dockerfile:58]: I think it should be possible to pass the build argument with clusterctl package `package=./cmd/clusterctl` and make the output binary an argument with a default to `manager` like this:
```
ARG outbin=manager
```
Then in Makefile you can:
```
DOCKER_BUILDKIT=1 docker build --build-arg builder_image=$(GO_CONTAINER_IMAGE) --build-arg goproxy=$(GOPROXY) --build-arg ARCH=$(ARCH) --build-arg package=./cmd/clusterctl --build-arg outbin=clusterctl --build-arg ldflags="$(LDFLAGS)" . -t $(CLUSTERCTL_IMG)-$(ARCH):$(TAG)
```

This way we will avoid including any other binaries in the clusterctl image.

Comment (sachinkumarsingh092) [reply Dockerfile:58]: That's great! Thanks. 
Also, do we have to make changes to push the image to `gcr.io` or will that be done as part of the release?

File: Dockerfile
Comment (ykakarap) [inline Dockerfile:64]:
```suggestion
```

I dont think you need a second `RUN` section here. This will lead to building both the manager and the clusterctl binaries (all though one will fail). You can re-purpose the existing `RUN` command so that the same command can be used to build the manager and clusterctl.  Changing the existing `RUN` to something like this should be enough: 
```
RUN --mount=type=cache,target=/root/.cache/go-build \
    --mount=type=cache,target=/go/pkg/mod \
    CGO_ENABLED=0 GOOS=linux GOARCH=${ARCH} \
    go build -trimpath -ldflags "${ldflags} -extldflags '-static'" \
    -o ${outbin} ${package}
```

Note that you will also have to replace "manager" in l.68 and l.71 with `${outbin}`.

**Review by ykakarap — CHANGES_REQUESTED** [review]

File: Dockerfile
Comment (sachinkumarsingh092) [reply Dockerfile]: Thanks. Made the required changes.

**Comment by sachinkumarsingh092** [comment]
I'm getting this error in the build logs:
```
#17 0.718 cmd/clusterctl/client/cluster/cert_manager.go:53:13: pattern assets/cert-manager-test-resources.yaml: no matching files found
#17 0.718 cmd/clusterctl/config/embedded_manifest.go:24:12: pattern manifest/clusterctl-api.yaml: no matching files found
#17 ERROR: executor failed running [/bin/sh -c CGO_ENABLED=0 GOOS=linux GOARCH=${ARCH}     go build -trimpath -ldflags "${ldflags} -extldflags '-static'"     -o ${outbin} ${package}]: exit code: 1  
```

But executing the following locally works:
```
CGO_ENABLED=0 GOOS=linux GOARCH=amd64     go build -trimpath -ldflags "-X 'sigs.k8s.io/cluster-api/version.buildDate=2022-04-14T12:54:47Z' -X 'sigs.k8s.io/cluster-api/version.gitCommit=a9b7087d8efd84e01b3d602b92bd24246ca80c51' -X 'sigs.k8s.io/cluster-api/version.gitTreeState=clean' -X 'sigs.k8s.io/cluster-api/version.gitMajor=1' -X 'sigs.k8s.io/cluster-api/version.gitMinor=1' -X 'sigs.k8s.io/cluster-api/version.gitVersion=v1.1.0-beta.2.388-a9b7087d8efd84' -X 'sigs.k8s.io/cluster-api/version.gitReleaseCommit=da03567d620d57b08fb9bd40e32de0da45b60391' -extldflags '-static'"     -o clusterctl ./cmd/clusterctl
```

This is just the expanded RUN section in the Dockerfile. A thing to note is that I don't build it inside a container with `--mount` option as in the RUN section, but I don't think that should be problem, should it be?

**Comment by ykakarap** [comment]
> I'm getting this error in the build logs:
> 
> ```
> #17 0.718 cmd/clusterctl/client/cluster/cert_manager.go:53:13: pattern assets/cert-manager-test-resources.yaml: no matching files found
> #17 0.718 cmd/clusterctl/config/embedded_manifest.go:24:12: pattern manifest/clusterctl-api.yaml: no matching files found
> #17 ERROR: executor failed running [/bin/sh -c CGO_ENABLED=0 GOOS=linux GOARCH=${ARCH}     go build -trimpath -ldflags "${ldflags} -extldflags '-static'"     -o ${outbin} ${package}]: exit code: 1  
> ```
> 
> But the files do exist, at least locally.
> 
> But executing the following locally works:
> 
> ```
> CGO_ENABLED=0 GOOS=linux GOARCH=amd64     go build -trimpath -ldflags "-X 'sigs.k8s.io/cluster-api/version.buildDate=2022-04-14T12:54:47Z' -X 'sigs.k8s.io/cluster-api/version.gitCommit=a9b7087d8efd84e01b3d602b92bd24246ca80c51' -X 'sigs.k8s.io/cluster-api/version.gitTreeState=clean' -X 'sigs.k8s.io/cluster-api/version.gitMajor=1' -X 'sigs.k8s.io/cluster-api/version.gitMinor=1' -X 'sigs.k8s.io/cluster-api/version.gitVersion=v1.1.0-beta.2.388-a9b7087d8efd84' -X 'sigs.k8s.io/cluster-api/version.gitReleaseCommit=da03567d620d57b08fb9bd40e32de0da45b60391' -extldflags '-static'"     -o clusterctl ./cmd/clusterctl
> ```
> 
> This is just the expanded RUN section in the Dockerfile. A thing to note is that I don't build it inside a container with `--mount` option as in the RUN section, but I don't think that should be problem, should it be?

This is because [all the yaml files are automatically ignored](https://github.com/kubernetes-sigs/cluster-api/blob/c7c1e1810e4909aa0f7a8e92e07767d39b1810b5/.dockerignore#L5) during docker builds and therefore these files are skipped when we perform `COPY`. 
We definitely need `assets/cert-manager-test-resources.yaml` and `manifest/clusterctl-api.yaml` when building clusterctl. Adding an exception for these specific files in `.dockerignore` should make sure that the files are copied over into the build container and that should be enough to fix the issue.

**Comment by sachinkumarsingh092** [comment]
The builds failed on this step:

```
STEP: Waiting for deployment capi-system/capi-controller-manager to be available


Failure [296.068 seconds]
[BeforeSuite] BeforeSuite 
/home/prow/go/src/sigs.k8s.io/cluster-api/test/e2e/e2e_suite_test.go:114 
```

On further inspection, it seems that `status.Available = True` condition was not satisfied due to which the test failed. Here is the status section of the Deployment:
```
     "status": {
      "observedGeneration": 1,
      "replicas": 1,
      "updatedReplicas": 1,
      "unavailableReplicas": 1,
      "conditions": [
        {
          "type": "Available",
          "status": "False",
          "lastUpdateTime": "2022-04-16T07:10:45Z",
          "lastTransitionTime": "2022-04-16T07:10:45Z",
          "reason": "MinimumReplicasUnavailable",
          "message": "Deployment does not have minimum availability."
        },
        {
          "type": "Progressing",
          "status": "True",
          "lastUpdateTime": "2022-04-16T07:10:45Z",
          "lastTransitionTime": "2022-04-16T07:10:45Z",
          "reason": "ReplicaSetUpdated",
          "message": "ReplicaSet \"capi-controller-manager-57f68646b8\" is progressing."
        }
      ]
    } 
```

Can it be due to a shortage of resources now that a new binary runs inside the container? These lines indicate something similar:
```
"reason": "MinimumReplicasUnavailable",
"message": "Deployment does not have minimum availability."
```

File: Dockerfile
Comment (ykakarap) [inline Dockerfile:49]:
```suggestion
ARG outbin="manager"
```

**Review by ykakarap — CHANGES_REQUESTED** [review]
The tests failed because the CAPI deployments never became ready. This is most likely because the docker image was built incorrectly. 

The default `outbin` should be `manager`, or else the target binary that is supposed to run is not properly defined and the container will fail to start leading the deployment to never be ready.

File: Dockerfile
Comment (sachinkumarsingh092) [reply Dockerfile:49]: But why won't passing `--build-arg outbin=manager` option work?

**Comment by sachinkumarsingh092** [comment]
It worked. Thanks a lot, @ykakarap. I also removed the `--build-arg outbin=manager` option in `docker-build-kubeadm-*` targets, but why wasn't this option working?

File: Dockerfile
Comment (ykakarap) [inline Dockerfile:65]: I am not a 100% sure how the clusterctl containers will be used but when we build the container like this we will not be able to run the clusterctl container as it fails with this: 
```
$ docker run --rm  gcr.io/kakaraparthy-devel/clusterctl-amd64:dev version

docker: Error response from daemon: failed to create shim: OCI runtime create failed: container_linux.go:380: starting container process caused: exec: "/bin/sh": stat /bin/sh: no such file or directory: unknown.
```

I think the problem is combination of using the build args to pass the binary name and the way the `ENTRYPOINT` is constructed. 

It might be cleaner to have a separate dockerfile for building clusterctl and a separate dockerfile that we can use for all the managers. 

Thoughts on this @fabriziopandini ?

**Comment by ykakarap** [comment]
> It worked. Thanks a lot, @ykakarap. I also removed the `--build-arg outbin=manager` option in `docker-build-kubeadm-*` targets, but why wasn't this option working?

That is because the issue was with the the core container (`docker-build-core`). The `docker-build-core` did not include the  `--build-arg outbin=manager` and therefore the container built as a result of this was invalid.

File: Makefile
Comment (sbueringer) [inline Makefile:445]: Please also adjust the docker push targets accordingly

File: Dockerfile
Comment (sbueringer) [reply Dockerfile]: I think if you add the push target accordingly it should be fine. This will publish to the staging registry and promotion to gcr.io during a release is done semi-manually via image promotion.

**Comment by sbueringer** [comment]
@CecileRobertMichon How does `make promote-images` pick up images for promotion?

**Comment by sbueringer** [comment]
@CecileRobertMichon Ah perfect. I was aware of the sort issue, but not about the image list (or I forgot it).

In my opinion we will push the clusterctl image to the staging repo and I think it makes sense to promote it to the production registry as well.

File: Dockerfile
Comment (ykakarap) [reply Dockerfile:50]: I believe this is because `ENTRYPOINT` does not support `ARG` values, however it does support using environment variables set using `ENV`.

Ref: https://docs.docker.com/engine/reference/builder/#environment-replacement

Comment (sbueringer) [reply Dockerfile:50]: Ah interesting

File: Dockerfile
Comment (ykakarap) [reply Dockerfile:65]: >I think using one Dockerfile should be fine. Why are we using exec instead of ["/${execbin}"] ?

I think this doesn't work too. I vaguely remember trying something like this and experiencing errors. IIRC it is because `[/manager]` is the Exec form and in this form a shell is not used to execute and therefore no environment variable substitutions are done because there is no shell to do it. 

May be @sachinkumarsingh092 can experiment with this and report any findings.

Comment (sachinkumarsingh092) [reply Dockerfile:65]: Yes. Another way to do this in exec for is ENTRYPOINT [ "sh", "-c", "$execbin" ], but this relies on the assumption that a `sh` shell is already present.
From the [docker docs](https://docs.docker.com/engine/reference/builder/#entrypoint):
```
Unlike the shell form, the exec form does not invoke a command shell. This means that normal shell processing does not happen. For example, ENTRYPOINT [ "echo", "$HOME" ] will not do variable substitution on $HOME. 
If you want shell processing then either use the shell form or execute a shell directly, for example: ENTRYPOINT [ "sh", "-c", "echo $HOME" ]. When using the exec form and executing a shell directly, as in the case for the shell form, it is the shell that is doing the environment variable expansion, not docker.
```

Comment (sbueringer) [reply Dockerfile:65]: So this mean we would have to use a base image with a shell just to make the name of the binary parameterizable?

If that's the case I think we should use a separate dockerfile.

Using the same binary name for controllers (aka manager) and clusterctl is probably not a good idea.

Comment (ykakarap) [reply Dockerfile:65]: I think the distroless image that we use does not come with a shell. Lets do with separate Dockerfiles.

File: Makefile
Comment (sbueringer) [inline Makefile:474]: Fyi, after this PR merges: https://github.com/kubernetes-sigs/cluster-api/pull/6452/files

You can rebase and add the clusterctl image to the list of images that should be promoted

**Comment by sachinkumarsingh092** [comment]
Hey all. Sorry for such a delayed response. I was testing if we could use another image such as `golang:alpine` to fit this case, but it didn't work. I had a few queries for the new Dockerfile:
- Where will this file reside? In `cmd/clusterctl` or in the root itself with a different name (`clusterctl.Dockerfile` maybe)?
- Should I use multistage builds (one for creating clusterctl and one for executing it with shell) or start with an image that has a shell in the beginning?

**Comment by ykakarap** [comment]
> Where will this file reside? In cmd/clusterctl or in the root itself with a different name (clusterctl.Dockerfile maybe)?

Lets have a Dockerfile in `cmd/clusterctl`.

Regarding using other base images: let's continue using the distroless image. I believe it is the recommended image by Kubernetes as well.

**Comment by sachinkumarsingh092** [comment]
We need `manager_pull_policy.yaml` and `manager_image_patch.yaml` in `cmd/clusterctl/config/default` for docker push targets. Are there any default kubebuilder markers to create them or should I create them manually?

**Comment by sbueringer** [comment]
> We need `clusterctl_pull_policy.yaml` and `clusterctl_image_patch.yaml` (akin to `manager_pull_policy.yaml` and `manager_image_patch.yaml` for kubeadm and core images) in `cmd/clusterctl/config/default` for docker push targets. Are there any default kubebuilder markers to create them or should I create them manually?

We don't need those files for clusterctl. You can omit the corresponding parts from the make targets. We only need them for other controllers because we are publishing "deployment YAMLs" for them.

File: .dockerignore
Comment (fabriziopandini) [inline .dockerignore:25]: Q: is this required?
This change doesn't seem related to this PR...

**Review by fabriziopandini — COMMENTED** [review]
Changes lgtm to me, but I'm not sure if all the points above are addressed; hopefully people involved in this PR can give feedback soon.

File: .dockerignore
Comment (sachinkumarsingh092) [reply .dockerignore:25]: This was done to fix this issue: https://github.com/kubernetes-sigs/cluster-api/pull/6405#issuecomment-1100418420

**Comment by sbueringer** [comment]
I'll try to take another look soon

File: cmd/clusterctl/Dockerfile
Comment (sbueringer) [inline cmd/clusterctl/Dockerfile:63]:
```suggestion
ENTRYPOINT ["/clusterctl"]
```

File: cmd/clusterctl/Dockerfile
Comment (sbueringer) [inline cmd/clusterctl/Dockerfile:38]: Q: Is there a relevant difference between:
```
COPY . .
```
and
```
COPY ./ ./
```

(I just diffed the Dockerfiles and I'm curious :). If there isn't let's use the same to keep the Dockerfiles as similar as posible.)

File: .dockerignore
Comment (sbueringer) [reply .dockerignore:25]: Sounds good to me

File: Makefile
Comment (sbueringer) [inline Makefile:474]:
```suggestion
	DOCKER_BUILDKIT=1 docker build --build-arg builder_image=$(GO_CONTAINER_IMAGE) --build-arg goproxy=$(GOPROXY) --build-arg ARCH=$(ARCH) --build-arg package=./cmd/clusterctl --build-arg ldflags="$(LDFLAGS)" -f ./cmd/clusterctl/Dockerfile . -t $(CLUSTERCTL_IMG)-$(ARCH):$(TAG)
```

**Review by sbueringer — COMMENTED** [review]
Thank you very much!
A few nits, otherwise lgtm from my side

Sorry for the slow response, too much going on right now..

File: cmd/clusterctl/Dockerfile
Comment (sachinkumarsingh092) [reply cmd/clusterctl/Dockerfile]: Don't know how this crept in. I tested on my local and it was working fine.
Made the correction.

**Comment by sbueringer** [comment]
Thank you very much.
/lgtm

If I'm correct this should be enough to build and push the clusterctl image e.g. in our nightly cloudbuild jobs.
It should also automatically promote the clusterctl image during release.

@CecileRobertMichon Would be great if you can take a look, I'm not sure if I missed something

**Comment by fabriziopandini** [comment]
/lgtm
/approve 
/hold
for @CecileRobertMichon pass

**Comment by vincepri** [comment]
@CecileRobertMichon are we ready to merge this PR?

**Comment by CecileRobertMichon** [comment]
Yes image promotion changes LGTM 

/hold cancel

**Pull request merged**
